src/rdkafka_msgset.h

```diff
@@ -30,6 +30,74 @@
 #define _RDKAFKA_MSGSET_H_
 
 
+
+/** 
+ * @struct rd_kafka_aborted_txns_t
+ * 
+ * @brief A collection of aborted transactions.
+ */
+typedef struct rd_kafka_aborted_txns_s {
+    rd_avl_t *avl;
+    /* Note: A list of nodes is maintained alongside
+     * the AVL tree to facilitate traversal.
+     */
+    rd_list_t *list;
+    int32_t cnt;
+} rd_kafka_aborted_txns_t;
+
+
+/**
+ * @brief Allocate memory for, and initialize a new
+ * rd_kafka_aborted_txns_t struct. 
+ */
+rd_kafka_aborted_txns_t *rd_kafka_aborted_txns_new (int32_t txn_cnt);
+
+
+/**
+ * @brief Free all resources associated with a
+ * rd_kafka_aborted_txns_t struct.
+ */
+void
+rd_kafka_aborted_txns_destroy (rd_kafka_aborted_txns_t *aborted_txns);
+
+
+/** 
+ * @brief Get the next aborted transaction start
+ * offset for the specified pid. Returns -1 if
+ * there is none.
+ */
+int64_t
+rd_kafka_aborted_txns_next_aborted_txn_offset_for_pid (rd_kafka_aborted_txns_t *aborted_txns,
+                                                         int64_t pid);
+
+
+/** 
+ * @brief Move to the next aborted transaction start
+ * offset for the specified pid.
+ */
+void
+rd_kafka_aborted_txns_move_to_next(rd_kafka_aborted_txns_t *aborted_txns,
```

Comment:
this is fighting muscle memory. did a pass through changing all of these, but missed one.

Reply:
or three

Reply:
the lookup without move_next is used in rd_kafka_msgset_reader_v2, so we need both. i don't think there's a good argument for efficiency here in spite of that as aborted transactions are not the normal case.

Reply:
what about `rd_kafka_aborted_txns_get_offset`?

Reply:
> . i don't think there's a good argument for efficiency here in spite of that as aborted transactions are not the normal case.

We did the worst-case napkin calculations, and there can be a large number of aborted transactions in a MessageSet - while it might not be normal we still need to handle it properly.

If we can easily avoid two lookups, we should.
And in this case there is just a single error check (`if (start_offset > hdr.offset)`) between the get and the pop, and that error check  may very well be done after the pop.

Every cycle counts

Reply:
yes, you are completely right, i was being delusional.